scripts/metrics/javadoc_eval.py

```diff
@@ -0,0 +1,181 @@
+#!/bin/python3
+
+# Evaluates the javadoc test results. It is required to execute javadoc.sh
+# prior to the execution of this script. This script parses the test
+# results saved in "build/results/javadoc" and converts them into the
+# "results.yml" format defined in the 
+# [documentation](https://github.com/Programmiermethoden/Deploy-to-Grading/blob/master/doc/design_document/d2g_procedure.md#format-der-result.yml).
+# The output is printed to the console. Make sure to execute the script
+# inside a task folder and that the task configuration defined in a
+# task.yml file was loaded correctly using the load_yaml scripts.
+#
+# Points are calculated by deducting a number of points specified by the
+# %TASKNAME%_METRICS_JAVADOC_DEDUCTION_PER_ERROR env variable from a maximum
+# number of points specified by the %TASKNAME%_METRICS_JAVADOC_MAX_POINTS env
+# variable. The minimum number of points is zero.
+# This metric supports two different modes for calculating points. In the
+# first mode, every error results in a point deduction. In the second mode,
+# the errors are grouped by type and points are only deducted once per type.
+# The mode can be switched by setting %TASKNAME%_METRICS_JAVADOC_GROUP_ERRORS
+# env variable to true for the second mode.
+#
+# usage: javadoc_eval.py [taskname(optional)]
+#   Params:
+#   taskname    Name of the task used as prefix for the env variables.
+#               If no taskname is given, "task" is used as the default
+#               value.
+#
+# Error handling:
+# - Exits with an error code when either of the 
+#   %TASKNAME%_METRICS_JAVADOC_MAX_POINTS,
+#   %TASKNAME%_METRICS_JAVADOC_GROUP_ERRORS
+#   or %TASKNAME%_METRICS_JAVADOC_DEDUCTION_PER_ERROR environment variables
+#   are not set.
+#
+# This script is part of step 6 in the Deploy-to-Grading pipeline that
+# is executed for every task that includes the javadoc metric.
+#
+
+import os
+import sys
+import xml.etree.ElementTree as ET
+import yaml
+
+RESULT_FILE = "build/results/javadoc/main.xml"
+MAX_POINTS_ENV_KEY = "%s_METRICS_JAVADOC_MAX_POINTS"
+GROUP_ERRORS_ENV_KEY = "%s_METRICS_JAVADOC_GROUP_ERRORS"
+DEDUCTION_PER_ERROR_ENV_KEY = "%s_METRICS_JAVADOC_DEDUCTION_PER_ERROR"
+
+def _print_usage():
+    print("usage: javadoc_eval.py [taskname(optional)]")
+    print("       Make sure that javadoc.sh was executed prior to this")
```

Comment:
Gedankenexperiment: Was würde passieren, wenn wir den Gradle-Aufruf auch noch hier im Python-Skript mit erledigen würden (und damit das Shell-Skript rauswerfen würden)? 

Aktuell ist sind beide getrennt aufrufbar: Erstellen der Metrik vs. Auswerten der Metrik. Aber brauchen wir diese Flexibilität wirklich? Zumal es zu möglichen Fehlern führen kann?

Reply:
Mh. Der Unterschied wäre, dass wir statt zuerst alle Metriken auszuführen und danach alle Metriken auszuwerten, Metrik für Metrik ausführen und direkt danach auswerten würden, bevor wir die nächste Metrik anfassen. Im Endeffekt dürfte das keinen Unterschied machen.

Ich wollte aber auch noch schauen, wie ich auch ohne Anpassung der `XYZ_eval.py`-Skripte auf die Shell-Skripte verzichten kann und stattdessen z.B. direkt `./gradlew junit` aufrufen kann. Da wäre die jetzige Form besser, da wir damit keinen doppelten Code in den `XYZ_eval.py`-Skripten haben.